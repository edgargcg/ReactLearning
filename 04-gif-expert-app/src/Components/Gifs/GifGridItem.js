import React from 'react'
import PropTypes from 'prop-types'

export const GifGridItem = ({ title, url}) => {
    return (
        <div className="images__item animate__animated animate__bounce">
            <h5 className="image__title">
                {title}
            </h5>
            <img 
                src={url} 
                alt={title} 
                className="image__source" 
            />
        </div>
    )
}

GifGridItem.propTypes = {
    url: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired
}