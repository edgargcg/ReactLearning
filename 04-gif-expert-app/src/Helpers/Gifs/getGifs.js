export const getGifs = async(category) => {
    const url = `https://api.giphy.com/v1/gifs/search?q=${encodeURI(category)}&limit=5&api_key=cqDyquoPCZDl2iOHC4dDHzlf80orxefP`;
    const response = await fetch(url);
    const { data } = await response.json();
    const gifs = data.map(img => {
        return {
            id: img.id,
            title: img.title,
            url: img.images?.downsized_medium.url
        }
    });

    return gifs;
}
