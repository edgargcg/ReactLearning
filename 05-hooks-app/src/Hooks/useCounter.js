import { useState } from 'react';

export const useCounter = ( initialValue = 10 ) => {
    const [counter, setCounter] = useState(initialValue)

    const increment = ( factor = 1 ) => {
        setCounter( counter + factor );
    }

    const decrement = ( factor = 1 ) => {
        setCounter( counter - factor );
    }

    const reset = ( resetValue = initialValue ) => {
        setCounter(resetValue);
    }

    return {
        counter,
        reset,
        decrement,
        increment,
    };
}
