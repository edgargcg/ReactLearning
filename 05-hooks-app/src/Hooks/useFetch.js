import { useEffect, useRef, useState } from "react"

export const useFetch = ( url ) => {
    const [state, setState] = useState({
        data: null,
        error: null,
        loading: true
    });

    const isMounted = useRef(true);

    useEffect(() => {
        return () => {
            isMounted.current = false;
        }
    }, [])

    useEffect(() => {
        
        setState(
            {
                data: null,
                error: null,
                loading: true
            }
        );

        fetch( url )
            .then( resp => resp.json()) 
            .then( data => {

                if (isMounted){
                    setState({
                        data: data,
                        error: null,
                        loading: false,
                    });
                }

            }).catch(error => {
                setState({
                    data: null,
                    error: error,
                    loading: false,
                });
            });

    }, [url]);

    return state;
}
