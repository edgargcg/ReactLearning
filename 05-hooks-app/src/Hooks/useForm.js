import { useState } from 'react'

export const useForm = ( initialState = {} ) => {

    const [values, setValues] = useState(initialState);

    const handleInputChange = (e) => {
        const { name, value } = e.target;

        setValues(
            {
                ...values,
                [name]: value
            }
        );
    }

    return[
        values,
        handleInputChange
    ];

}
