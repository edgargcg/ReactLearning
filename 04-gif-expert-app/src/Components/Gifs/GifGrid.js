import React, {  } from 'react'
import { useFetchGifs } from '../../Hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';
import PropTypes from 'prop-types';

export const GifGrid = ({ category }) => {
    const { data: images, loading } = useFetchGifs(category);

    return (
        <div>
            <h3>
                {category}
            </h3>
            {loading && <p>Loading</p> }
            <div className="images">
                {
                    images.map(
                        img => 
                        (
                            <GifGridItem 
                                {...img} 
                                key={img.id} 
                            />
                        )
                    )
                }
            </div>
        </div>
    )
}


GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}