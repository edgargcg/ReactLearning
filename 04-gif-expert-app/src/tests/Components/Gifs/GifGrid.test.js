import { shallow } from 'enzyme';
import { GifGrid } from '../../../Components/Gifs/GifGrid';
import { useFetchGifs } from '../../../Hooks/useFetchGifs';

jest.mock('../../../Hooks/useFetchGifs');

describe('Test on <GifGrid />', () => {
    const category = 'Batman';
    
    test('should render <GifGrid />', () => {
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrid category={category} />);
        expect(wrapper).toMatchSnapshot();
    })

    test('should render items on load images with useFetchGifs', () => {
        const gifs = [
            {
                id: 'ABC',
                url: 'https://jola.com',
                title: 'Cualquier cosa'
            },
            {
                id: '123',
                url: 'https://jola.com',
                title: 'Cualquier cosa'
            }
        ];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow(<GifGrid category={category} />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
    })
        
    
})
