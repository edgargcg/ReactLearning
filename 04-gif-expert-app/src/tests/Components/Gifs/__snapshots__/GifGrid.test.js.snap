// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <GifGrid /> should render <GifGrid /> 1`] = `
<div>
  <h3>
    Batman
  </h3>
  <p>
    Loading
  </p>
  <div
    className="images"
  />
</div>
`;

exports[`Test on <GifGrid /> should render items on load images with useFetchGifs 1`] = `
<div>
  <h3>
    Batman
  </h3>
  <div
    className="images"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualquier cosa"
      url="https://jola.com"
    />
    <GifGridItem
      id="123"
      key="123"
      title="Cualquier cosa"
      url="https://jola.com"
    />
  </div>
</div>
`;
